docs: reorganize and refresh readme

sskeirik opened the pull request:
Fixes: #1734 

I wrote this after walking through the K Framework build process with a client and noticed that there were some dependencies that were listed in the "short version" but not the "long version."

I updated the "long version" section to move the list of the dependencies to the top and then put the installation notes at the bottom. I think this reads a bit more nicely --- it gives you an upfront picture of the entire process without getting bogged in the details at first.

EDIT: I also made a few other changes:

- I added a quick description of what the K Framework is.
- I added a table of contents
- I added some notes about supported system configurations
- I reorganized the section on building the OCaml backend slightly
- I added some notes about building the K Framework on macOS/brew systems
- I renamed and reorganized the "Installing on a Fresh Windows Subsystem for Linux"

Questions:

- Do we officially support 32 bit systems? It's not clear to me.
- Is the readme getting to be too long and verbose? Should we break it out into separate documents?

Thoughts?

bmmoore commented:
I haven't tried installing OCaml on windows to test this changes, but other than the comments above it looks good.

bmmoore commented:
Ok, looks good.

Commit:
docs: update prereq installation instructions

Commit:
docs: make pkg-config an explicit dependency

Reason: the LLVM backend needs it

Commit:
docs: update supported system configurations section

Commit:
docs: tweak the quick start guide for ubuntu bionic

Commit:
docs: update the build and install guide section

Commit:
docs: add toc to the readme

Commit:
docs: add what-is-the-k-framework-about to readme

sskeirik commented:
@ehildenb @dwightguth @ttuegel This PR is for updating the main README for the reasons stated in the original description. I'm not sure who should review this.

That said, do you guys have info about my pending questions above?
Any feedback in general would be appreciated.

ehildenb commented on README.md:
````diff
@@ -199,6 +266,72 @@ mkShell {
+    ```
+
+    For VirtualBox or virtual machine software users, you will need to open
+    the VirtualBox or virtual machine application and follow the graphical
````
Likely some of this should go into `k-distribution/INSTALL.md` instead, and you should link there from here. Anything to do with getting K standing up on a given platform from our packaged releases should go into `INSTALL.md`. This guide should focus on (i) from-source builds for developers, and (ii) how developers can get examples running so they can begin working on K.

ehildenb commented:
@sskeirik it looks mostly OK so far. I would emphasize that packaged installation instructions should go into `k-distribution/INSTALL.md`, and that the `README.md` is intended _for developers_ to be able to begin working on K, and doesn't make it into any user-facing generated documentation.

Commit:
move quick start guide/touch up README

Commit:
fix URL in README

sskeirik commented:
@ehildenb I updated the README as suggested and moved the quick start guide section into the INSTALL.md file.

The quick start could be further parameterized so that it applies to multiple systems --- let me know if you are interested in that.

Besides that one possible issue, I think this is ready for re-review.

ehildenb commented on README.md:
```diff
@@ -7,17 +7,95 @@ copyright: Copyright (c) 2010-2020 K Team. All Rights Reserved.
+
+*   Windows (any architecture)
+*   macOS (M1)
+
```
This is packaging information and should remain in k-distribution/INSTALL.md. We do not want to duplicate information whenever possible, rather link to the one true source.

ehildenb commented on README.md:
```diff
@@ -7,17 +7,95 @@ copyright: Copyright (c) 2010-2020 K Team. All Rights Reserved.
+
+*   Windows 8 and older (x86-64) users can install
+    [VirtualBox](https://www.virtualbox.org) and then:
+    1.   create a new virtual machine (VM);
```
We do not need to include instructions in the K README about setting up virtual development environments. Users can google that on their own. You can include a single sentence or two saying that they will need to set it up, but instruct them to look up their own guides.

ehildenb commented on k-distribution/INSTALL.md:
```diff
@@ -181,3 +181,60 @@ eval $(opam config env)
 above commands with `export OPAMROOT=/path/to/opam/root`.
+
+Quick Start Guide for Ubuntu Bionic/Focal
+-----------------------------------------
```
```suggestion
Quick Start Guide
-----------------
```

ehildenb commented:
Why is there a haskell backend update in this PR? Please revert that.

Commit:
docs: reorganize INSTALL.md

Commit:
docs: make supported configuration info link to INSTALL.md

Commit:
docs: reorganize installation notes section in README

ehildenb commented:
Please also remove `k-distribution/README.md`, and change the `src/main/assembly` XML file to reflect that (use the top-level README).

ehildenb commented:
We do _not_ support 32-bit machines.

Commit:
docs: README update date

Commit:
docs: delete old k-distribution README

Commit:
build: update maven assembly to use top-level README

Commit:
docs: fix typos in INSTALL.md

Commit:
README: fix URLs/add comment on 32 bit systems

sskeirik commented:
@ehildenb @dwightguth The requested changes have been made and this should be ready for re-review.

Note that README.md file currently links to the INSTALL.md file on the master branch (so it will be out of date if you click on it), but will work correctly once this PR is merged. This is helpful because relative URLs break during the package assembly process (unless we do some sort of URL rewriting which seems overkill).

sskeirik committed:
Merge branch 'master' into readme_update

sskeirik committed:
Merge branch 'master' into readme_update

sskeirik committed:
Merge branch 'master' into readme_update
